### PR TITLE
issue-45

### PR DESCRIPTION
**Pull request by stephen-pl**
---

<details open="true"><summary>Generated summary (powered by <a href="https://app.graphite.dev">Graphite</a>)</summary>

> # Pull Request Description
> 
> ## TL;DR
> This pull request implements the logic to run the sp filtering job. The job will run periodically to figure out which sp are eligible for a tenant based on the eligibility criteria specified by the tenant. .
> 
> ## What changed
> - Schema changed from text to jsonb for clause_value
> - Reduced scope of comparison operators 
> - Code to filter sp for a tenant 
> 
> ## How to test
> 1. Compile and run the code.
> 2. Verify that the code compiles successfully.
> 3. Test the functionality of the comparison operators.
> 4. Verify that the code behaves as expected.
> 
> ---

### CODE REVIEW

**Comment by stephen-pl** [comment]
Current dependencies on/for this PR:
* `main`
  * **PR #27** <a href="https://app.graphite.dev/github/pr/data-preservation-programs/spade-tenant/27?utm_source=stack-comment-icon" target="_blank">Graphite</a>  👈

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/data-preservation-programs/spade-tenant/27?utm_source=stack-comment).

File: jobs/spFilteringJob.go [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+				if !compare(clause.ClauseValue, getField(&sp, clause.ClauseAttribute), clause.ClauseOperator) {
+					isValid = false
+					break
+				}
+			}
+
+			if isValid {
+				// which states should be not allow bumps back to eligible
+				spID, err := strconv.Atoi(sp.Provider[1:])
+				if err != nil {
```
Comment (stephen-pl) [inline jobs/spFilteringJob.go:96]: we should probably just skip a bad provider id

File: jobs/spFilteringJob.go [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+	case db.ExcludedFrom:
+		return !contains(tenantSetting, spSetting)
+	default:
+		return false
+	}
+}
+
+func contains(tenantSetting []string, spSetting string) bool {
+	for _, setting := range tenantSetting {
+		if strings.Contains(strings.ToLower(setting), strings.ToLower(spSetting)) {
```
Comment (stephen-pl) [inline jobs/spFilteringJob.go:129]: this will need to be === since one value could be a substring.

File: db/schema.go [hunk 3]
```diff
@@ -103,8 +104,29 @@ type TenantSPEligibilityClauses struct {
 	TenantID        ID                 `json:"tenant_id" gorm:"primaryKey"`
 	ClauseAttribute string             `json:"attribute" gorm:"primaryKey"`
 	ClauseOperator  ComparisonOperator `json:"operator" gorm:"type:comparison_operator;not null"`
-	ClauseValue     string             `json:"value" gorm:"not null"`
+	ClauseValue     StringSlice        `json:"value" gorm:"type:JSON;not null"`
 }
+
+type StringSlice []string
```
Comment (jcace) [inline db/schema.go:110]: can you add a comment just explaining why we had to do this (no support in Gorm for array types)

File: jobs/spFilteringJob.go (hunk 1 above)
Comment (jcace) [reply jobs/spFilteringJob.go:96]: agree, log an error message and `continue` .

File: jobs/spFilteringJob.go [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+			for _, clause := range clauses {
+				if !compare(clause.ClauseValue, getField(&sp, clause.ClauseAttribute), clause.ClauseOperator) {
+					isValid = false
+					break
+				}
+			}
+
+			if isValid {
+				// which states should be not allow bumps back to eligible
+				spID, err := strconv.Atoi(sp.Provider[1:])
```
Comment (jcace) [inline jobs/spFilteringJob.go:95]: add a comment about why `[1:]`

File: jobs/spFilteringJob.go [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+				if err != nil {
+					panic(err)
+				}
+
+				if database.Model(&db.TenantsSPs{}).Updates(&db.TenantsSPs{TenantID: tenantId, SPID: db.ID(spID), TenantSpState: db.TenantSpStateEligible}).RowsAffected == 0 {
```
Comment (jcace) [inline jobs/spFilteringJob.go:100]: you also need to create the `&db.SP{}` entry in the `SPs` table, if it doesn't already exist

Comment (jcace) [inline jobs/spFilteringJob.go:100]: hmm, this needs to intelligently handle a transition to the new `disabled` state, i.e, if the SP was previously eligible/pending/active and then became ineligible (ex, they moved location), then the `TenantsSPs` state needs to be set to `disabled` (this new state is in my PR [here](https://github.com/data-preservation-programs/spade-tenant/pull/25/files#diff-84e606ea1f6e7818300439a7188c129bd6f36eb3aae0adccd19b3cc4d29b81b1R145)). So we could do it as part of another PR or just add the `Disabled` state in your PR here too.

File: jobs/spFilteringJob.go [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+
+		tenantClauseMap[clause.TenantID] = append(tenantClauseMap[clause.TenantID], clause)
+	}
+
+	for tenantId, clauses := range tenantClauseMap {
+		for _, sp := range sps {
+			isValid := true
+
+			for _, clause := range clauses {
+				if !compare(clause.ClauseValue, getField(&sp, clause.ClauseAttribute), clause.ClauseOperator) {
```
Comment (jcace) [inline jobs/spFilteringJob.go:87]: do you think it would be possible to write some unit tests for the business logic of this filtering job? 

The best would be if we can extract the logic (iteration/comparison) apart from the DB stuff into a separate function, and then run some unit tests against it for a variety of clauses/SPs. At minimum some unit tests of the `compare/contains/getField` would be good.

File: jobs/spFilteringJob.go [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+				if database.Model(&db.TenantsSPs{}).Updates(&db.TenantsSPs{TenantID: tenantId, SPID: db.ID(spID), TenantSpState: db.TenantSpStateEligible}).RowsAffected == 0 {
+					database.Model(&db.TenantsSPs{}).Create(&db.TenantsSPs{TenantID: tenantId, SPID: db.ID(spID), TenantSpState: db.TenantSpStateEligible})
+				}
+			}
+		}
+	}
+
+}
+
+// TODO @jcace its important that we force the field names a user can select to be the field of the object or we need to maintain an equivalent mapping somewhere. Lets discuss sometime this week.
```
Comment (jcace) [inline jobs/spFilteringJob.go:109]: yes - for v1 here we just need to make sure we hardcode the exact values from the MongoDB somewhere so that it won't grab the wrong field

File: jobs/spFilteringJob.go (hunk 5 above)
Comment (stephen-pl) [reply jobs/spFilteringJob.go:100]: I think there are some subtleties here we need to think through. ill schedule a meeting sometime to discuss the state transitions. I have a feeling itll be faster than back and forth on a pr.

File: jobs/spFilteringJob.go (hunk 7 above)
Comment (stephen-pl) [reply jobs/spFilteringJob.go:109]: Agreed.

File: jobs/spFilteringJob.go (hunk 6 above)
Comment (stephen-pl) [reply jobs/spFilteringJob.go:87]: Absolutely. This was more of a wip pr to flush out any issues with the model and business logic.

**Comment by stephen-pl** [comment]
disable flow should be working as expected now. Going to focus on tests next

File: jobs/spFilteringJob.go [hunk 8]
```diff
@@ -0,0 +1,167 @@
...
+	AgentRegion    string    `bson:"agentRegion"`
+	Date           time.Time `bson:"date"`
+	LatencyMs      float64   `bson:"latencyMs"`
+	Multiaddr      string    `bson:"multiaddr"`
+	TestId         string    `bson:"testId"`
+}
+
+func main() {
+
+	sps := setUpEnvAndFetchSpDataFromMongo()
```
Comment (jcace) [inline jobs/spFilteringJob.go:37]: nit: move `godoteenv.Load()` to the beginning of `main`, then rename this function to simply `fetchSpData()`

File: jobs/spFilteringJob.go [hunk 9]
```diff
@@ -0,0 +1,167 @@
...
+		}
+	}
+}
+
+func setUpEnvAndFetchSpDataFromMongo() []SP {
+	godotenv.Load()
+	serverAPI := options.ServerAPI(options.ServerAPIVersion1)
+	opts := options.Client().ApplyURI(os.Getenv("MONGODB_URI")).SetServerAPIOptions(serverAPI)
+
+	client, err := mongo.Connect(context.TODO(), opts)
```
Comment (jcace) [inline jobs/spFilteringJob.go:63]: make this function take a `ctx context.Context` from the caller, then pass that in to this `mongo.Connect()` function. Instantiate the `ctx := context.TODO()` at the top level of `main()` instead

File: jobs/spFilteringJob.go [hunk 10]
```diff
@@ -0,0 +1,167 @@
...
+		panic(err)
+	}
+
+	defer func() {
+		if err = client.Disconnect(context.TODO()); err != nil {
+			panic(err)
+		}
+	}()
+
+	collection := client.Database(os.Getenv("MONGODB_DATABASE")).Collection(os.Getenv("MONGODB_SP_COLLECTION"))
```
Comment (jcace) [inline jobs/spFilteringJob.go:74]: what happens if one or both of these is undefined?

File: jobs/spFilteringJob.go [hunk 11]
```diff
@@ -0,0 +1,167 @@
...
+			log.Fatal(err)
+		}
+
+		sps = append(sps, sp)
+	}
+
+	return sps
+}
+
+func isValidClause(clauses []db.TenantSPEligibilityClauses, sp SP) bool {
```
Comment (jcace) [inline jobs/spFilteringJob.go:91]: hmm, I'm a bit confused by the naming `isValid` and `isValidClause`, could this be more accurately named as `isEligible` and `verifyEligibility` ?

File: jobs/spFilteringJob_test.go [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+
+	if !compare(us, "US", db.IncludedIn) {
+		t.Errorf("Expected %q to contain US", us)
+	}
+
+	if compare(us, "US", db.ExcludedFrom) {
+		t.Errorf("Expected %q to be excluded from contain US", us)
+	}
+
+	countries := []string{"AH", "CA", "US", "DE", "MX"}
```
Comment (jcace) [inline jobs/spFilteringJob_test.go:20]: not part of this PR specifically, but we do need to think about / document what are the valid country/region/city codes that one would find in Kentik, so we can populate those correctly when applying criteria for a tenant. 

Initially it could be as simple as a spreadsheet with a big list of valid values, to make sure we pick from that list when populating the `EligibilityClauses`

File: jobs/spFilteringJob.go (hunk 10 above)
Comment (stephen-pl) [reply jobs/spFilteringJob.go:74]: im pretty sure it panics but ill double check

File: jobs/spFilteringJob_test.go (hunk 12 above)
Comment (stephen-pl) [reply jobs/spFilteringJob_test.go:20]: agreed. our validation story needs work

File: jobs/spFilteringJob.go (hunk 10 above)
Comment (stephen-pl) [reply jobs/spFilteringJob.go:74]: Yep panics with runtime error invalid memory address

File: jobs/spFilteringJob.go (hunk 11 above)
Comment (stephen-pl) [reply jobs/spFilteringJob.go:91]: good call out I think that is clearer as well

File: jobs/spFilteringJob_test.go (hunk 12 above)
Comment (stephen-pl) [reply jobs/spFilteringJob_test.go:20]: Here is a [task](https://github.com/orgs/data-preservation-programs/projects/12/views/5?sliceBy%5Bvalue%5D=stephen-pl&pane=issue&itemId=53551792) to track that work

**Review by jcace — APPROVED** [review]
LGTM
